src/core/irc_client.ml

```diff
@@ -265,17 +274,20 @@ module Make(Io: Irc_transport.IO) = struct
 
   let connect
       ?username ?(mode=0) ?(realname="irc-client")
-      ?password ?(sasl=true) ?config ~addr ~port ~nick () =
+      ?password ?(sasl=Some `Plain) ?config ~addr ~port ~nick () =
     Io.open_socket ?config addr port >>= (fun sock ->
       let connection = mk_connection_ sock in
 
       let cap_end = ref false in
       begin
-        match username, password with
-        | Some user, Some password when sasl ->
+        match username, password, sasl with
+        | Some user, Some password, Some `Plain ->
+          cap_end := true;
+          send_auth_sasl_plain ~connection ~user ~password
+        | Some user, _, Some `External ->
           cap_end := true;
-          send_auth_sasl ~connection ~user ~password
-        | _, Some password -> send_pass ~connection ~password
+          send_auth_sasl_external ~connection ~user
+        | _, Some password, None -> send_pass ~connection ~password
         | _ -> return ()
```

Comment:
Maybe we should match on etc `None, _, Some `Plain | _, None, Some `Plain` and issue an error instead of silently skipping SASL authentication. What do you think, @johnelse ?

Reply:
Since this is going to be a breaking change anyway, we could go even further and do something like

```
type auth = {
    password : string;
    sasl : [`Plain | `External] option
}
```

and have `connect` take an `auth option`. That way, there's no way of requesting any kind of SASL without also supplying a password.

Reply:
I think that makes sense. Can I suggest that `auth` also carries a username field? It is not always the case that you want to authenticate with the same username as the IRC user.

The SASL external mechanism does not use the password for anything. Maybe the type should have a different shape.